### ISSUE TITLE
SolidJS: Use a solid store for the messages array for better referential differentiation in `<For/>`

### ISSUE THREAD

**Issue by spiritledsoftware** [vercel/ai#2002]
### Description

Because the messages array is a signal (well really a signal inside a resource), when using the signal inside a `<For />` any update to the array causes a complete rerender. I have a `<For />` iterating over the array inside a scrollable div and every time the array updates, the scroll position resets.

### Code example

```tsx
const {
    input,
    setInput,
    handleSubmit,
    reload,
    messages,
    error,
    isLoading
  } = useChat();

const [startOfMessagesRef, setStartOfMessagesRef] = createSignal<HTMLDivElement>();
const startOfMessagesVisible = createVisibilityObserver({ initialValue: true })(
  startOfMessagesRef
);

const [endOfMessagesRef, setEndOfMessagesRef] = createSignal<HTMLDivElement>();
const endOfMessagesVisible = createVisibilityObserver()(endOfMessagesRef);
createEffect(
  on(endOfMessagesVisible, (endOfMessagesVisible) => {
    if (endOfMessagesVisible && messagesQuery.hasNextPage && !messagesQuery.isFetchingNextPage) {
      messagesQuery.fetchNextPage();
    }
  })
);

return <div class="flex grow flex-col-reverse items-center space-y-2 overflow-y-auto whitespace-pre-wrap border-b">
  <div ref={setStartOfMessagesRef} />
    <For
      each={messages()?.toReversed() || []}
      fallback={
        <div class="flex h-full w-full items-center justify-center">
          <H5>No messages yet</H5>
        </div>
      }
    >
  {(message, idx) => (
    <div data-index={idx()} class="flex w-full max-w-2xl flex-col">
      <Message message={message} />
      <Separator orientation="horizontal" class="w-full" />
    </div>
  )}
  </For>
  <div ref={setEndOfMessagesRef} />
</div>
```

### Additional context
I am trying to achieve infinite scroll on the messages as well using tanstack query, but whenever I update the messages array to the query results, the update causes the scroll position to reset all the way to the bottom! Not ideal! Also, during a response stream, the scroll position will constantly reset as well if the user is trying to scroll around during the response.

In discussion on the SoldJS discord, it sounds like array would be better stored in a SolidJS store object so that the `reconcile` function can be used, which automatically handles diffing and merging based on a specific key.

**Comment by spiritledsoftware** [comment]
It seems like this may be a problem with the underlying `solid-swr-store` package… I have also created an issue in that repo, but not sure how well maintained that library is considering the last commit was a year ago!

**Comment by spiritledsoftware** [comment]
Created an this issue on the other repo: https://github.com/lxsmnsyc/swr-store/issues/3

**Comment by spiritledsoftware** [comment]
This was completed in #2095

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Bug**  
Confidence: **High**

### Rationale
This is a concrete failure in the existing SolidJS integration: message updates lose item identity and trigger disruptive full rerenders and scroll resets. The proposed store/reconciliation change is a fix for that behavior.

### Evidence
- Updating the SolidJS `useChat` messages array causes `<For>` to rerender all entries and resets the scroll position.
- The reset also occurs repeatedly while a response is streaming, disrupting user scrolling.
- The reporter identified the underlying `solid-swr-store` behavior and noted that the fix was completed in PR #2095.
